app/components/file-viewer/index.jsx

```diff
@@ -20,39 +21,52 @@ const VIEWERS = {
   image: ImageViewer,
   text: TextViewer,
   video: VideoPlayer,
-  audio: AudioPlayer
+  audio: AudioPlayer,
+  canvas: CanvasViewer
 };
 
 function subjectViewerSelector(props) {
   if (Array.isArray(props.type)) {
     if (props.type.includes('audio')) {
       return VIEWERS.audio;
     }
-    // ... add outher here if neccessary
+    // ... add other here if neccessary
   }
   return VIEWERS[props.type] || DefaultViewer;
 }
 
 function FileViewer(props) {
   const Viewer = subjectViewerSelector(props);
-
+  const viewerProps = {
+    className: props.className,
+    style: props.style,
+    src: props.src,
+    type: props.type,
+    format: props.format,
+    frame: props.frame,
+    onLoad: props.onLoad,
+    onFocus: props.onFocus,
+    onBlur: props.onBlur
+  };
+  if (props.type === 'canvas') {
+    Object.assign(
+      viewerProps,
+      {
+        annotations: JSON.parse(JSON.stringify(props.classification.annotations)),
+        subject: props.subject,
+        viewBoxDimensions: props.viewBoxDimensions,
+        workflow: props.workflow
+      }
+    );
+  }
   return (
-    <Viewer
-      className={props.className}
-      style={props.style}
-      src={props.src}
-      type={props.type}
-      format={props.format}
-      frame={props.frame}
-      onLoad={props.onLoad}
-      onFocus={props.onFocus}
-      onBlur={props.onBlur}
-    />
+    <Viewer {...viewerProps} />
   );
 }
 
 FileViewer.propTypes = {
   className: React.PropTypes.string,
+  classification: React.PropTypes.object,
```

Comment:
Remember that the file viewer will also be used to view subjects on Talk, where classifications and workflows aren't available.

I'd advise against making any changes to this component that tie it to the classifier.

Reply:
I'm not sure where to send the annotation/zoom information to in a way it can be used then. Do you have any suggestions?

Reply:
I thought you'd decided to pass that info down via the frame annotator components?

Reply:
It's going to the file-viewer from frame-annotator and frame-viewer, then through to canvas-viewer. Could add a default value for classification, annotation and subject (no longer passing workflow) in file-viewer?

Reply:
Can you remind me why the classification and subject are being passed down? At this level in the component tree, you're dealing with a single frame from a subject, so I'm not really getting why the subject still needs to be available.

If you already have the annotation as a prop, what are you using the classification for?

Sorry for the questions, I'm just very confused by the intent of this component at the moment.

Reply:
Actually, would it make more sense to generate the model somewhere around here, then pass it down as a prop instead of the subject?

Reply:
Primary motivation behind this component is to allow in-browser modelling, as is currently done in a hacky way by https://zooniverse.org/projects/tingard/galaxy-builder.

Entire classification is being passed down as some projects like galaxy-builder rely on the whole annotations list to create a model, as each task is for a different part of the model (disk/bar/spiral). Tried doing this in one huge combo task and UX was horrible.

Annotation prop is being used to check whether the most recent annotation has changed, though I could just as easily use `classification.annotations[classification.annotations.length -1]`

Subject is needed as there are many possible tasks that canvas-viewer could do under the umbrella of modelling, and I need a way to identify which model *this* frame of *this* subject should be using, hence storing it in metadata.

Reply:
In general I'd need the canvas context (2d or webgl) to render, and while assembling a function which could be called with the context would be okay for 2d contexts, it would be super expensive to repeatedly create and bind webgl rendering commands. I can give it a go but doubt it'd work well.
Is there any movement behind storing annotations in redux? Then I wouldn't need to thread anything through other components and could just grab from the store using a wrapper. Doesn't address the problem of canvas viewer being classification independent though...

Reply:
Have a look at #4277 and #4272 for where we currently ate with annotations. We want to use a store for annotations in the new classifier, which we're about to start building, but the current proposal is that the subject viewer in the new classifier will be read-only ie. it will be designed to render subject frames but won't have any knowledge of annotations.

Reply:
> Entire classification is being passed down as some projects like galaxy-builder rely on the whole annotations list to create a model, as each task is for a different part of the model (disk/bar/spiral). Tried doing this in one huge combo task and UX was horrible.

The model only needs to be updated when the current annotation changes, though? The values for previous tasks will be fixed so recalculating the mode for those seems wasteful. Is it possible to store the model and update with annotation changes, rather than creating it from scratch each time? Then pass the model down to a canvas that renders it.

Reply:
That does make sense. I suppose the problem is that what I'm proposing results in the classification becoming part of the subject, which goes against that ethos.

I suppose that would be possible, but then you'd still be passing through what is effectively a compiled version of the annotation, breaking the whole annotation-free file-viewer plan.

How about somehow separating into a "dynamic-file-viewer" and a static one?

Reply:
Sounds good. Your idea of a compiled version of the annotations sounds good to me. Would the annotations be compiled into something that could be stored in Redux, then read and rendered by the viewer?

Reply:
I could probably organise things into a basic redux compatible object, would need some fiddling to check.
As far as I can tell you'd need to send all annotations through as canvas-viewer will unmount and re-mount when frames are switched, so if the first task is done on an image subject and then the next requires the canvas, canvas-viewer wouldn't know the first annotation to render, which it might need in the general case.
Did you mean the idea of a dynamic-file-viewer sounds good?

Reply:
I was thinking the viewer should be a purely functional component that renders its props, and all the heavy calculation work should be done higher in the tree, stored in state and passed down as props to the viewer. That avoids you redoing all your work when the frame changes, instead reading your data from a prop or, eventually, a data store.

Reply:
Also, avoid coding for the general case unless you actually need it. There's a few bits of code in the classifier that try to accommodate cases which, it turns out, aren't actually needed in real projects.